packages/@aws-cdk/aws-appmesh/lib/gateway-route-spec.ts

```diff
@@ -1,32 +1,395 @@
+  /**
+   * GatewayRoute CFN configuration for gRPC gateway route match.
+   */
+  readonly requestMatch: CfnGatewayRoute.GrpcGatewayRouteMatchProperty;
+}
+
+/**
+ * Optional properties for gRPC gateway route match other than the host name.
+ */
+export interface GrpcGatewayRouteHostnameMatchOptions {
+  /**
+   * The gateway route metadata to be matched on
+   *
+   * @default - no match on metadata
+   */
+  readonly metadata?: GrpcMetadataMatch[];
+
+  /**
+   * The fully qualified domain name for the service to match from the request.
+   *
+   * @default - no match on service name
+   */
+  readonly serviceName?: string;
+}
+
+/**
+ * Optional properties for gRPC gateway route match other than the metadata.
+ */
+export interface GrpcGatewayRouteMetadataMatchOptions {
+  /**
+   * The gateway route host name to be matched on.
+   *
+   * @default - no match on host name
+   */
+  readonly hostname?: GatewayRouteHostname;
+
+  /**
+   * The fully qualified domain name for the service to match from the request.
+   *
+   * @default - no match on service name
+   */
+  readonly serviceName?: string;
+}
+
+/**
+ * Optional properties for gRPC gateway route match other than the service name.
+ */
+export interface GrpcGatewayRouteServiceNameMatchOptions {
+  /**
+   * The gateway route host name to be matched on.
+   *
+   * @default - no match on host name
+   */
+  readonly hostname?: GatewayRouteHostname;
+
+  /**
+   * The gateway route metadata to be matched on.
+   *
+   * @default - no match on metadata
+   */
+  readonly metadata?: GrpcMetadataMatch[];
+}
+
+/**
+ * The criterion for determining a request match for this GatewayRoute.
+ */
+export abstract class GrpcGatewayRouteMatch {
+  /**
+   * Create host name based gRPC gateway route match.
+   *
+   * @param hostname The gateway route host name to be matched on
+   * @param options Optional properties to define gRPC gateway route match
+   */
+  public static hostname(hostname: GatewayRouteHostname, options?: GrpcGatewayRouteHostnameMatchOptions): GrpcGatewayRouteMatch {
+    return new GrpcGatewayRouteHostnameMatchImpl(hostname, options);
+  }
+
+  /**
+   * Create metadata based gRPC gateway route match.
+   *
+   * @param metadata An object that represents the data to match from the request.
+   * @param options Optional properties to define gRPC gateway route match
+   */
+  public static metadata(metadata: GrpcMetadataMatch[], options?: GrpcGatewayRouteMetadataMatchOptions): GrpcGatewayRouteMatch {
+    return new GrpcGatewayRouteMetadataMatchImpl(metadata, options);
+  }
+
+  /**
+   * Create service name based gRPC gateway route match.
+   *
+   * @param serviceName The fully qualified domain name for the service to match from the request
+   * @param options Optional properties to define gRPC gateway route match
+   */
+  public static serviceName(serviceName: string, options?: GrpcGatewayRouteServiceNameMatchOptions): GrpcGatewayRouteMatch {
```

Comment:
> Which means all 3 options allow passing all 3 types of parameters!
>
> And that one of them is required? Because if that's the case, then let's just go with the interface, and check at runtime that one of them has been passed...

Yes, you are correct. Requirement is to choose at least one option but may choose all. Will go with the interface + runtime check approach.

> Also, I see that HttpGatewayRouteMatch is still an interface. Given that, it seems wrong to make GrpcGatewayRouteMatch a union-like class.

I left `HttpGatewayRouteMatch` an an interface because this has a default value to set (*set `prefix` to `/`) whereas `GrpcGatewayRouteMatch` does not have any default value.